src/data_provider.rs

```diff
@@ -200,84 +233,105 @@ pub mod imp {
             if let Some((host, port)) = addr {
                 self.set_service_address(host, port);
             }
-            
+
             self.establish_connection();
         }
 
         pub fn disconnect(&self) {
             println!("Disconnect request...");
-            self.stop_signal.store(true, Ordering::SeqCst);
-            if let Some(handle) = self.handle.borrow_mut().take() {
-                handle.join().unwrap();
+            self.task_runner.take();
+            self.handle.take();
+            if let Some(joinhandle) = self.join_handle.take() {
+                joinhandle.join().unwrap();
                 //drop(self.admin_client.clone());
                 println!("Client thread stopped!");
             }
         }
 
-        pub fn start_vm(&self, name: String) {
-            let admin_client = self.admin_client.clone();
-            thread::spawn(move || {//not sure it is really needed
-                    Runtime::new().unwrap().block_on(async move {
-                    //there is only one name in my disposal
-                    if let Err(error) = admin_client.read().unwrap().start(name.clone(), None, Vec::<String>::new()).await {
-                        println!("Start request error {error}");
-                    }
-                    else {
-                        println!("Start request sent");
-                    };
-                })
+        fn client_cmd<F: FnOnce(Result<(), String>) -> () + 'static>(&self, task: Task, cb: F) {
+            let (res_tx, res_rx) = async_channel::bounded(1);
+            let Some(tr) = self.task_runner.borrow().as_ref().cloned() else {
+                cb(Err("Not connected to admin".into()));
+                return;
+            };
+            glib::spawn_future_local(async move {
+                // On error res_tx is dropped and res_rx.recv() will fail below
+                let _ = tr.send((task, res_tx)).await;
+                let res = match res_rx.recv().await {
+                    Ok(res) => res,
+                    Err(err) => Err(err.to_string()),
+                };
+                cb(res);
             });
         }
+        
+        pub fn start_vm(&self, name: String) {
+            self.client_cmd(
+                adminclient!(|client| client.start("".to_owned(), Some(name), vec![]).await),
```

Comment:
the args handling was changed, the first must be the name of app or service and the second is VM (which can be omitted for now - None).

Reply:
Changed.